packages/core/src/shared/utilities/map.ts

```diff
@@ -0,0 +1,98 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from '../logger'
+
+/**
+ * Simply maps a Key to an Object. Everything is done in memory, allowing for atomic updates.
+ *
+ * There is a common use case for this class, so look to reuse this when possible.
+ * Benefits, instead of making your own:
+ * - A key does not have to be a string, {@link MemoryMap.asKey} enables this
+ * - Implements most of the boilerplate
+ */
+export abstract class MemoryMap<Key, Value = { [key: string]: any }> implements MapSync<Key, Value> {
+    private readonly data: { [key: string]: Value | undefined } = {}
+
+    exists(key: Key): boolean {
+        return this.data[this.asKey(key)] !== undefined
+    }
+
+    /**
+     * IMPORTANT: If getting a key without having {@link set()} a value, the {@link default} will be returned.
+     * Use {@link exists()} to check for existence before calling get(), if necessary.
+     */
+    get(key: Key): Value {
+        const actualKey = this.asKey(key)
+        const result = this.data[actualKey]
+        return result ?? this.default
+    }
+
+    set(key: Key, data: Partial<Value>): void {
+        const currentData = this.get(key)
+        this.data[this.asKey(key)] = { ...currentData, ...data }
+    }
+
+    clear(key: Key, reason: string): void {
+        delete this.data[this.asKey(key)]
+        getLogger().debug(`${this.name}: cleared cache, key: ${JSON.stringify(key)}, reason: ${reason}`)
+    }
+
+    /**
+     * Converts the user-provided key to a string so that it can be used
+     * as an object key.
+     */
+    protected abstract asKey(key: Key): string
+
+    /**
+     * The name of the implementation, for logging purposes
+     */
+    protected abstract get name(): string
+
+    /**
+     * The default value returned from {@link get}() when {@link exists}() is false
+     */
+    protected abstract get default(): Value
+}
+
+/**
+ * A synchronous version of a Map of Maps.
+ * - This allows for atomic updates as it is not async.
+ * - They Key does not need to be a string
+ */
+export interface MapSync<Key, ReturnObject = { [key: string]: any }> {
+    /**
```

Comment:
What is this abstraction communicating? "not async" implies there is an async variant (where?).

Is there a reason this doesn't derive from the ES `Map` interface?

Reply:
should this live in `cacheUtils.ts` ? it seems very related.

Reply:
> There are some differences with the Map interface, 

is that difference big enough to lose the discoverability/intuitiveness gained by implementing `Map` ? 

> such as not allowing partial values in `set()` that wouldn't allow it to be used.

I am wondering about this. Is it just a typing issue? I think we can work around it.